pkg/storage/chunk/client/ibmcloud/trusted_profile_authentication_provider.go

```diff
@@ -0,0 +1,159 @@
+package ibmcloud
+
+import (
+	"os"
+
+	"github.com/IBM/go-sdk-core/v5/core"
+	"github.com/IBM/ibm-cos-sdk-go/aws/awserr"
+	"github.com/IBM/ibm-cos-sdk-go/aws/credentials"
+	"github.com/IBM/ibm-cos-sdk-go/aws/credentials/ibmiam/token"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/loki/pkg/util/log"
+	"github.com/pkg/errors"
+)
+
+const (
+	trustedProfileProviderName = "TrustedProfileProviderNameIBM"
+	tokenType                  = "Bearer"
+)
+
+// TrustedProfileProvider Struct
+// This implements Provider interface from https://github.com/IBM/ibm-cos-sdk-go
+type TrustedProfileProvider struct {
+	// Name of Provider
+	providerName string
+
+	// Type of Provider - SharedCred, SharedConfig, etc.
+	providerType string
+
+	// Authenticator implements an IAM-based authentication schema
+	authenticator *core.ContainerAuthenticator
+
+	// Error
+	ErrorStatus error
+}
+
+// NewTrustedProfileProvider allows the creation of a custom IBM IAM Provider
+// Parameters:
+//
+//	Provider Name
+//	AWS Config
+//	Trusted Profile Name
+//	Trusted Profile ID
+//	Compute Resource Token File Path
+//	IBM IAM Authentication Server Endpoint
+//
+// Returns:
+//
+//	TrustedProfileProvider
+func NewTrustedProfileProvider(providerName string, trustedProfileName,
+	trustedProfileID, crTokenFilePath, authEndpoint string) *TrustedProfileProvider {
+	provider := new(TrustedProfileProvider)
+
+	provider.providerName = providerName
+	provider.providerType = "oauth"
+
+	if trustedProfileName == "" && trustedProfileID == "" {
+		provider.ErrorStatus = awserr.New("trustedProfileNotFound", "either Trusted profile name or id must be provided", nil)
+		level.Debug(log.Logger).Log("msg", provider.ErrorStatus)
+
+		return provider
+	}
+
+	if crTokenFilePath == "" {
+		provider.ErrorStatus = awserr.New("crTokenFilePathEmpty", "must supply cr token file path", nil)
+		level.Debug(log.Logger).Log("msg", provider.ErrorStatus)
+
+		return provider
+	} else {
+		if _, err := os.Stat(crTokenFilePath); errors.Is(err, os.ErrNotExist) {
+			provider.ErrorStatus = awserr.New("crTokenFileNotFound", "no such file", err)
+			level.Debug(log.Logger).Log("msg", "no such file", "err", err)
```

Comment:
which file? if you get the error in the logs, will we be able to identify which file we are talking about?